### ISSUE TITLE
Single sample inference

### ISSUE THREAD

**Issue by maricante** [PIX2NVS/NVS2Graph#8]
I have trained a 3-class classifier as described in your paper by using Train.py with batch_size=64. The trained model performs wonderfully on larger batches (acc > 90%), but accuracy drops significantly when performing inference on smaller batches. For batch_size=1 accuracy is ~50%.

Do you have any ideas as to what could be the culprit here? Ideally, I would like the model to perform well on single samples. Thanks!

**Comment by PIX2NVS** [comment]
Hi, Maricante, thanks for your interests.
Did you try inference on a larger batch? I do not think the bach size in the inference can influence the results.
If you only trained a 3-class classifier using the ASL-DVS, you are supposed to obtain higher accuracy.

**Comment by maricante** [comment]
Thanks for the reply.

I have trained the classifier using the [RoShamBo dataset](https://docs.google.com/document/d/e/2PACX-1vTNWYgwyhrutBu5GpUSLXC4xSHzBbcZreoj0ljE837m9Uk5FjYymdviBJ5rz-f2R96RHrGfiroHZRoH/pub), which I have converted to graphs as described in your paper. I have tested my conversion script on the ASL-DVS dataset and the resulting graphs are similar to what I get with RoShamBo, so I believe it works well. The RoShamBo dataset is, however, a bit noisier than the ASL-DVS, so I believe this is causing the lower accuracy.

I have not changed the architecture of the network itself in the Train.py script, only the cluster sizes in the pooling layers to accommodate the resolution of DAVIS240c (as described in the paper) and the number of exit nodes in the fc2 layer to 3.

With these settings, testing on different batch sizes does yield different results for me:
- batch size 64 (same as training batch) gives accuracy of 91%
- batch size 1 gives accuracy of 55%
- batch size 128 gives accuracy of 8%

Do your results on the ASL-DVS dataset stay the same regardless of testing batch size?

Searching around, I have found that [batchnorm might be the cause](https://discuss.pytorch.org/t/model-eval-gives-incorrect-loss-for-model-with-batchnorm-layers/7561). I am currently testing the potential solutions from this thread.

I apologize if I am missing something obvious, as I am very new to this area.

**Comment by PIX2NVS** [comment]
Thanks for providing details.

I did not try the inference using different batch size as I think the batch size in inference does not influence results much. I do not have exactly correct ideas about why the gap of results is so large when you are using different when using different batch size. 

But I guess there may be some problem in the input of model or the method how to bath the graphs. Double-check if the .mat files you generate are the same as mine in the repo, including data type, class(unit8, single...). 

As RoShamBo is not a complex dataset with only three classes, the model may be overfitting. Try using fewer layers and fewer neurons to see what you can get.  Or as you mentioned, try to delete the batch norm layer to see the results.

Please let me know if you figure out the problem. I am curious about this problem :)

**Comment by maricante** [comment]
Thanks for the tips.

Indeed, my labels weren't formatted as uint8. However, after reformatting them to uint8, the issue still remains.

I will try out the rest of your suggestions and let you know if I manage to solve the problem.
